util/strong_random.py

```diff
@@ -0,0 +1,137 @@
+        """
+
+        self.buffer = np.load(input_file)
+        self.empty = False
+        self.next = 0
+
+    def size(self):
+        """ Size of the buffer.
+        """
+        return np.size(self.buffer)
+
+    def printstatus(self):
+        """Print the status of the buffer and the number of remaining bytes.
+
+        May be useful for testing and debugging.
+        """
+
+        print(self.buffer)
+        if self.empty:
+            print('Buffer is empty.')
+        else:
+            print("Buffer remaining:", len(self.buffer), "bytes.")
+
+    def isempty(self):
+        """Checks if the buffer is empty.
+        """
+        return self.empty
+
+    def fetchbyte(self):
+        """Fetch a random byte from the buffer.
+
+        Used by getrandbits().
+        Fetches a random byte from a buffer and updates the self.empty status.
+        If the buffer is already empty, issues a warning and defaults to
+        pseudo-random numbers from random.py.
+        """
+
+        if self.empty:
+            _warn('Buffer empty, defaulting to random.py', UserWarning)
+            random_byte = rnd.getrandbits(8)
+            return (random_byte)
+        else:
+            random_byte = self.buffer[0]
+            self.buffer = np.delete(self.buffer, 0)
+            self.empty = (self.buffer.size == 0)
+            return (random_byte)
+
+    def getrandbits(self, n_bits):
+        """Fetch n_bits random bits from a buffer.
+
+        Always fetches a whole number of bytes. Unused bits are discarded.
+        """
+
+        bitsleft = n_bits
+        R = 0
+        while (bitsleft > 0):
+            random_byte = self.fetchbyte()
+            if (bitsleft > 8):
+                R = (R << 8) | random_byte
+                bitsleft = bitsleft - 8
+            else:
+                random_byte >>= 8 - bitsleft
+                R = (R << bitsleft) | random_byte
+                bitsleft = 0
+        return (R)
+
+    def randbellow(self, n):
+        """Generate a random integer smaller than n.
+
+        Uses getrandbits() to generate enough random bits. Repeats until the
+        resulting number is smaller than n. The probability of this is more
+        than 50% in each loop iteration.
+        The amount of consumed randomness is not fixed.
+        """
+        length_bits = _ceil(_log(n, 2))
+        R = self.getrandbits(length_bits)
+        while R >= n:
+            R = self.getrandbits(length_bits)
+        return (R)
+
+    def shuffle(self, x):
+        """Shuffle a list x.
+
+        Each possible permutation of x is generated with equal probability.
+        """
+
+        length_list = len(x)
+        for i in reversed(range(length_list)):
+            j = self.randbellow(i + 1)
+            x[i], x[j] = x[j], x[i]
+
+
+# ----------------------------------------------------------------------
+# Create one instance, and export its methods as module-level functions.
+# The functions share state across all uses.
```

Comment:
What does usage of this class look like, exactly? Will the entropy buffers be loaded from more than one place? I'm asking because I'm concerned the shared state could lead to entropy being repeated, for instance if one user loads the entropy buffer into the global state and retrieves some bytes, and then another user loads the same buffer into the global state and retrieves the same bytes.

Reply:
The intention is to load the entropy buffer only once and to generate all netlist constants. This is similar to how we are using `random.seed()` function in `topgen.py` at the moment.
We indeed need to be careful to not load the same file twice, but I'm not sure if there is a way around it.

Reply:
Hmm, maybe the shared instance actually comes in handy here. If we expect only one load during the whole runtime, would it be reasonable to add a check to the `load` function that errors out if a file has already been loaded?

Reply:
It can't hurt to be extra careful. I've modified the `srnd.load()` to raise an error in case that the user attempts to load the same file twice. We can easily remove it later if needed, but I don't think it can cause problems.

Reply:
Sounds good.

Reply:
This sounds good to me and would catch any user error.

As of now, topgen would be the first place where we are going to integrate this - but there are two more scripts that use randomness, and these are the life cycle state generator and the OTP memory map generator.

They are currently not part of the top-gen invocation, but should probably be in the future (via IpGen, since OTP is technically a top-level dependent IP).

We can deal with this later however, once we integrated this class into topgen.